acvm-repo/brillig_vm/src/arithmetic.rs

```diff
@@ -107,89 +112,71 @@ pub(crate) fn evaluate_binary_int_op<F: AcirField>(
+        BinaryIntOp::LessThan => !lhs & rhs,
+        BinaryIntOp::LessThanEquals => lhs <= rhs,
         BinaryIntOp::And => lhs & rhs,
         BinaryIntOp::Or => lhs | rhs,
         BinaryIntOp::Xor => lhs ^ rhs,
-        BinaryIntOp::Shl => {
-            if rhs >= 128 {
-                0
-            } else {
-                lhs.wrapping_shl(rhs as u32)
-            }
-        }
-        BinaryIntOp::Shr => {
-            if rhs >= 128 {
-                0
+        BinaryIntOp::Shl | BinaryIntOp::Shr => {
+            if rhs {
+                false
             } else {
-                lhs.wrapping_shr(rhs as u32)
+                lhs
             }
         }
     };
     Ok(result)
 }
 
-fn evaluate_binary_int_op_generic(
+fn evaluate_binary_int_op_num<
+    T: PrimInt + AsPrimitive<usize> + From<bool> + WrappingAdd + WrappingSub + WrappingMul,
+>(
     op: &BinaryIntOp,
-    lhs: u128,
-    rhs: u128,
-    bit_size: IntegerBitSize,
-) -> Result<u128, BrilligArithmeticError> {
-    let bit_size: u32 = bit_size.into();
-    let bit_modulo = 1 << bit_size;
+    lhs: T,
+    rhs: T,
+    num_bits: usize,
+) -> Result<T, BrilligArithmeticError> {
     let result = match op {
-        // Perform addition, subtraction, and multiplication, applying a modulo operation to keep the result within the bit size.
-        BinaryIntOp::Add => (lhs + rhs) % bit_modulo,
-        BinaryIntOp::Sub => (bit_modulo + lhs - rhs) % bit_modulo,
-        BinaryIntOp::Mul => (lhs * rhs) % bit_modulo,
-        // Perform unsigned division using the modulo operation on a and b.
-        BinaryIntOp::Div => {
-            if rhs == 0 {
-                return Err(BrilligArithmeticError::DivisionByZero);
-            } else {
-                lhs / rhs
-            }
-        }
-        // Perform a == operation, returning 0 or 1
-        BinaryIntOp::Equals => (lhs == rhs) as u128,
-        // Perform a < operation, returning 0 or 1
-        BinaryIntOp::LessThan => (lhs < rhs) as u128,
-        // Perform a <= operation, returning 0 or 1
-        BinaryIntOp::LessThanEquals => (lhs <= rhs) as u128,
-        // Perform bitwise AND, OR, XOR, left shift, and right shift operations, applying a modulo operation to keep the result within the bit size.
+        BinaryIntOp::Add => lhs.wrapping_add(&rhs),
+        BinaryIntOp::Sub => lhs.wrapping_sub(&rhs),
+        BinaryIntOp::Mul => lhs.wrapping_mul(&rhs),
+        BinaryIntOp::Div => lhs.checked_div(&rhs).ok_or(BrilligArithmeticError::DivisionByZero)?,
+        BinaryIntOp::Equals => (lhs == rhs).into(),
+        BinaryIntOp::LessThan => (lhs < rhs).into(),
+        BinaryIntOp::LessThanEquals => (lhs <= rhs).into(),
         BinaryIntOp::And => lhs & rhs,
         BinaryIntOp::Or => lhs | rhs,
         BinaryIntOp::Xor => lhs ^ rhs,
         BinaryIntOp::Shl => {
-            if rhs >= (bit_size as u128) {
-                0
+            let rhs_usize = rhs.as_();
+            if rhs_usize >= num_bits {
```

Comment:
Will this also overflow if `lhs_num_bits + rhs_usize >= num_bits`? (Same for `Shr` case)

Reply:
This is just to cover the handling injected by rust where it disallows shifting by more than the type's bit size. It's also unchanged from the previous implementation.